### PR TITLE
Add Serenata lsp client

### PR DESCRIPTION
**Pull request by Sasanidas**
Client support for the PHP language server [Serenata](https://gitlab.com/Serenata/Serenata)

### CODE REVIEW

**Comment by kiennq** [comment]
Unrelated to this, but can we group all `php` language server into a single file?
Rust already doing that, right now we have an `php` language server under `lsp-clients`, another one in `lsp-intelliphense` and this can be a third one under another file.
While I'm believe on you only need to pay what you get concept, grouping all language server for a language under a same file is actually easier for discovery, both from maintainer POV and user (who don't want to spend much time reading though help page)

File: lsp-serenata.el [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  "Define serenata start function, it requires a PORT."
+  `(, lsp-serenata-server-path
+    "-u" ,(number-to-string port)))
+
+
+(lsp-register-client
+ (make-lsp-client
+  :new-connection (lsp-tcp-connection 'lsp-serenata-server-start-fun)
+  :major-modes '(php-mode)
+  :priority -1
```
Comment (yyoncho) [inline lsp-serenata.el:53]: the priority should not be the same as the other php servers - how do you compare seranata to the others?

Comment (kiennq) [reply lsp-serenata.el:53]: We should really group all `php` related language server into one file, it's also easier to compare the priority too

Comment (yyoncho) [reply lsp-serenata.el:53]: > We should really group all `php` related language server into one file, it's also easier to compare the priority too

Let's keep this as part of the different PR and eventually decide if we want to follow that. Splitting into multiple files as advantages too - e. g. you might skip loading the file via lsp-client-packages.

Comment (Sasanidas) [reply lsp-serenata.el:53]: > the priority should not be the same as the other php servers - how do you compare seranata to the others?
Serenata is less powerfull than intelphense or php-language-server, it still doesn't have find-references or workspace-symbols.
 It is the less mature of the 3, so maybe the lowest priority

**Comment by yyoncho** [comment]
LGTM, thank you @Sasanidas 

@kiennq are you fine with merging as it is and then discuss separately how to structure the language-specific features?

@ericdallo - would you take a look at the docs? Also, is it possible to check how the docs look like before merging a PR?

File: docs/lsp-clients.json [hunk 2]
```diff
@@ -351,6 +351,14 @@
...
     "debugger": "Not available"
   },
+  {
+    "name": "serenata",
+    "full-name": "PHP",
+    "server-name": "Serenata",
+    "server-url": "https://gitlab.com/Serenata/Serenata",
+    "installation-url": "https://gitlab.com/Serenata/Serenata/-/releases",
+    "debugger": "Yes"
+  },
```
Comment (ericdallo) [inline docs/lsp-clients.json:361]: Could you move this to above the other PHP servers?

File: lsp-serenata.el [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+;;; Code:
+
+(require 'lsp-mode)
+
+(defgroup lsp-serenata nil
+  "LSP support for the PHP programming language, using serenata"
+  :group 'lsp-mode
+  :link '(url-link "https://gitlab.com/Serenata/Serenata"))
+
+
```
Comment (ericdallo) [inline lsp-serenata.el:34]:
```suggestion
```

File: lsp-serenata.el [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+(defcustom lsp-serenata-server-path
+  "serenata.phar"
+  "Path to the Serenata Language Server phar file.
+It can be downloaded from https://gitlab.com/Serenata/Serenata/-/releases."
+  :group 'lsp-serenata
+  :type 'file)
+
+
```
Comment (ericdallo) [inline lsp-serenata.el:42]:
```suggestion
```

File: lsp-serenata.el [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+(defun lsp-serenata-server-start-fun (port)
+  "Define serenata start function, it requires a PORT."
+  `(, lsp-serenata-server-path
+    "-u" ,(number-to-string port)))
+
+
```
Comment (ericdallo) [inline lsp-serenata.el:48]:
```suggestion
```

File: docs/lsp-clients.json [hunk 6]
```diff
@@ -351,6 +351,14 @@
     "installation": "install.packages(“languageserver”)",
     "debugger": "Not available"
   },
+  {
+    "name": "serenata",
+    "full-name": "PHP",
```
Comment (ericdallo) [inline docs/lsp-clients.json:356]:
```suggestion
    "full-name": "PHP (Serenata)",
```

**Comment by kiennq** [comment]
> @kiennq are you fine with merging as it is and then discuss separately how to structure the language-specific features?

Yes, it's fine to me

File: lsp-serenata.el [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  "serenata.phar"
+  "Path to the Serenata Language Server phar file.
+It can be downloaded from https://gitlab.com/Serenata/Serenata/-/releases."
+  :group 'lsp-serenata
+  :type 'file)
+
+
+(defun lsp-serenata-server-start-fun (port)
+  "Define serenata start function, it requires a PORT."
+  `(, lsp-serenata-server-path
```
Comment (kiennq) [inline lsp-serenata.el:44]: nit: no space
```suggestion
  `(,lsp-serenata-server-path
```

**Comment by Sasanidas** [comment]
> LGTM, thank you @Sasanidas
> 
> @kiennq are you fine with merging as it is and then discuss separately how to structure the language-specific features?
> 
> @ericdallo - would you take a look at the docs? Also, is it possible to check how the docs look like before merging a PR?

I'm planing of contribute the lsp [paslm](https://psalm.dev/docs/running_psalm/language_server/) server in the future, so an unification approach would be interesting.

File: lsp-serenata.el [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+;; lsp-serenata client
+
+;;; Code:
+
+(require 'lsp-mode)
+
+(defgroup lsp-serenata nil
+  "LSP support for the PHP programming language, using serenata"
+  :group 'lsp-mode
```
Comment (ericdallo) [inline lsp-serenata.el:31]: @yyoncho do we should add `:package-version '(lsp-mode . "6.4")` here?

**Comment by yyoncho** [comment]
Thank you for contributing to `lsp-mode` and congrats for your first `lsp-mode` PR! :tada:

**Comment by erikhu** [comment]
ey guy, how can use serenata in emacs, i tried in many ways but i can't get it still. i don't understand the documentation in lsp page, i downloaded the phar but always the emacs try to use the iph.

**Comment by yyoncho** [comment]
The point is that `iph` is with higher priority. You have several options, one of them is to disable `iph` with

``` emacs-lisp
(setq lsp-disabled-clients '(iph))
```

**Comment by erikhu** [comment]
thank you so much @yyoncho  i almost got it, but now it show me this

LSP :: serenata::tcp has exited (exited abnormally with code 255)
Server serenata:276165 exited with status closed. Do you want to restart it? (y or n) y

that's a progress for me, because a have hours trying to figure out how to resolve

**Comment by yyoncho** [comment]
Now check  the corresponding stderr buffer. Probably something serenata
specific is off.

erikhu writes:

> thank you so much @yyoncho  i almost got it, but now it show me this
>
> LSP :: serenata::tcp has exited (exited abnormally with code 255)
> Server serenata:276165 exited with status closed. Do you want to restart it? (y or n) y
>
> that's a progress for me, because a have hours trying to figure out how to resolve

**Pull request merged**